src/commands/logging/logger.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2017 Oursky Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import sqlFormatter from 'sql-formatter';
+
+import { LogColorizer } from './coloring';
+
+export type Logger = (logData: any) => string;
+
+export type ColorizedLogger =
+  (logColorizer: LogColorizer, logData: any) => string;
+
+export const keyvalueLogger: ColorizedLogger = (logColorizer, logData) => {
+  const logLine = Object.keys(logData)
+    .map((eachKey) => ({key: eachKey, value: logData[eachKey]}))
+    .map(({key, value}) => `${key}=${JSON.stringify(value)}`)
+    .join(' ');
+  const detailsLog = (({ sql, error }) => {
+    if (sql) {
+      // Remove the redundant spaces after '$'
+      return sqlFormatter.format(sql).replace(/\$\s+/g, '$');
+    }
+
+    if (error) {
+      return error;
```

Comment:
not sure the type of `error`, if it is an Error and this function returns a string, perhaps this need type cast to string? (question)

Reply:
The type of `error` should be `string`. The following is a sample of structured log containing `error` key:

```
{
    "error": "Traceback (most recent call last):\n  File \"/root/.local/lib/python3.5/site-packages/skygear/transmitter/common.py\", line 43, in wrapper\n    return dict(result=f(self, *args, **kwargs))\n  File \"/root/.local/lib/python3.5/site-packages/skygear/transmitter/common.py\", line 134, in call_handler\n    return self.handler(func, param)\n  File \"/root/.local/lib/python3.5/site-packages/skygear/transmitter/common.py\", line 146, in handler\n    response = func(request)\n  File \"__init__.py\", line 27, in throw_exception\n    raise Exception('Some Exception')\nException: Some Exception",
    "level": "error",
    "logger": "skygear.transmitter.common",
    "msg": "Error occurred processing request",
    "tag": "plugin",
    "time": "2018-07-10T08:35:46Z"
}
```